### ISSUE TITLE
Replace existential JwsHeader method arguments in SigningKeyResolverAdapter with something sane

### ISSUE THREAD

**Issue by dwalend** [jwtk/jjwt#83]
SigningKeyResolverAdapter uses existential - raw - JwsHeader in its arguments.

My immediate problem is that I can't override those methods in Scala - no way to express the raw type in a way the compiler will swallow.

object ShrineSigningKeyResolver extends SigningKeyResolverAdapter {
    def resolveSigningKey(header: JwsHeader, claims: Claims):Key = { ... }}

can't be made to compile.

Long-term raw Java types are going to be a source of trouble in Java.

Thanks,

Dave

**Comment by dmitriyvolk** [comment]
David, did you manage to go around this problem somehow? Thank you!

**Comment by dwalend** [comment]
I hacked around it with (the only) lump of Java code in SHRINE. It'll likely work for Java through JDK10 and maybe JDK11, but the folks at Oracle do want to change existential types into a compile error. You'll want to fix this (maybe with higher kinded types, which Java can handle, but not gently) eventually. The easiest fix for you might be to remove the type parameter entirely. Groovy doesn't care.

Dave

package net.shrine.dashboard.jwtauth;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwsHeader;
import io.jsonwebtoken.SigningKeyResolverAdapter;

import java.security.Key;

/**
 * SigningKeyResolverBridge is in Java because JwsHeader takes a parameter, but SigningKeyResolverAdapter specifies it as a raw type.
 *
 * @author david
 * @since 1.21
 */
public class SigningKeyResolverBridge extends SigningKeyResolverAdapter {

    public Key resolveSigningKey(JwsHeader header, Claims claims) {
        String keyId = header.getKeyId();
        return KeySource.keyForString(keyId);
    }
}

**Comment by dmitriyvolk** [comment]
Thanks, Dave, I ended up doing something very similar. Was easier for me, too, because I didn't have to do anything general-purpose. I only had to provide an interface for resolving the key based on the issuer. 

Dmitriy

**Comment by lhazlewood** [comment]
What is the suggested fix for this issue?  Without a suggestion or PR, it doesn't seem like there is anything to be done?

**Comment by lhazlewood** [comment]
Is it to change the method signatures from these:

```java
Key resolveSigningKey(JwsHeader header, Claims claims);
Key resolveSigningKey(JwsHeader header, String plaintext);
```

to these:

```java
<T extends JwsHeader<T>> Key resolveSigningKey(JwsHeader<T> header, Claims claims);
<T extends JwsHeader<T>> Key resolveSigningKey(JwsHeader<T> header, String plaintext);
```

?

**Comment by dwalend** [comment]
That's probably the most type-safe option. Do you care about backwards compatibility?

**Comment by lhazlewood** [comment]
Yes, but to a slightly lesser degree while we're on 0.x.  Once we hit 1.0.0, we'll need to be very strict about semantic versioning.

That said, I don't think this will break anything - at least not in Java.  I made the method signature change (didn't change the Adapter at all), rebuilt the entire project, and all tests pass.

**Comment by dwalend** [comment]
No new warnings even? That's a surprise.

Thanks for making the fix,

Dave

**Comment by wengertj** [comment]
Are you still planning to change the method signature in `1.0.0` @lhazlewood ?
It would be great to ditch that `@SuppressWarnings( "rawtypes" )` annotation in my code 🙂

**Comment by lhazlewood** [comment]
Yes, this is still planned for 1.0.0, but we need to finish JWE and JSON (non-compact) serialization first.

**Comment by jdussouillez** [comment]
Any updates on this ?

**Comment by lhazlewood** [comment]
Same status as my last comment.  1.0 has been a long time coming, but we *are* actively working on it.

**Comment by lhazlewood** [comment]
This has been resolved with the new `Locator<Header>` and `JwtParserBuilder.keyLocator` concept that has been merged to `master` (`SigningKeyResolver is now deprecated).  This will be released with `0.12.0`
